blivet/zfcp.py

```diff
@@ -47,62 +49,218 @@ def logged_write_line_to_file(fn, value):
+                               "offline (%(e)s).")
+                             % {'devnum': self.devnum, 'e': e})
+
+    def _is_scsi_associated_with_fcp(self, fcphbasysfs, _fcpwwpnsysfs, _fcplunsysfs):
+        """Decide if the SCSI device with the provided SCSI attributes
+        corresponds to the zFCP device.
+
+        :returns: True or False
+        """
+
+        return fcphbasysfs == self.devnum
+
+    def online_device(self):
+        """Initialize the device and make its storage block device(s) ready to use.
+
+        :returns: True if success
+        :raises: ValueError if the device cannot be initialized
+        """
+
+        self._free_device()
+        self._set_zfcp_device_online()
+        return True
+
+    def offline_scsi_device(self):
+        """Find SCSI devices associated to the zFCP device and remove them from the system."""
+
+        # A list of existing SCSI devices in format Host:Bus:Target:Lun
+        scsi_devices = [f for f in os.listdir(scsidevsysfs) if re.search(r'^[0-9]+:[0-9]+:[0-9]+:[0-9]+$', f)]
+
+        scsi_device_found = False
+        for scsidev in scsi_devices:
+            fcpsysfs = os.path.join(scsidevsysfs, scsidev)
+
+            with open(os.path.join(fcpsysfs, "hba_id")) as f:
+                fcphbasysfs = f.readline().strip()
+            with open(os.path.join(fcpsysfs, "wwpn")) as f:
+                fcpwwpnsysfs = f.readline().strip()
+            with open(os.path.join(fcpsysfs, "fcp_lun")) as f:
+                fcplunsysfs = f.readline().strip()
+
+            if self._is_scsi_associated_with_fcp(fcphbasysfs, fcpwwpnsysfs, fcplunsysfs):
+                scsi_device_found = True
+                scsidel = os.path.join(scsidevsysfs, scsidev, "delete")
+                logged_write_line_to_file(scsidel, "1")
+                udev.settle()
+
+        if not scsi_device_found:
+            log.warning("No scsi device found to delete for zfcp %s", self)
+
+
+class ZFCPDevice(ZFCPDeviceBase):
+    """A class for zFCP devices that are not configured in NPIV mode. Such
+    devices have to be specified by a device number, WWPN and LUN.
+    """
+
+    def __init__(self, devnum, wwpn, fcplun):
+        super().__init__(devnum)
+
+        self.wwpn = blockdev.s390.zfcp_sanitize_wwpn_input(wwpn)
+        if not self.wwpn:
+            raise ValueError(_("You have not specified a worldwide port name or the name is invalid."))
+
+        self.fcplun = blockdev.s390.zfcp_sanitize_lun_input(fcplun)
+        if not self.fcplun:
+            raise ValueError(_("You have not specified a FCP LUN or the number is invalid."))
+
+    # Force str and unicode types in case any of the properties are unicode
+    def _to_string(self):
+        return "{} {} {}".format(self.devnum, self.wwpn, self.fcplun)
+
+    def _is_scsi_associated_with_fcp(self, fcphbasysfs, fcpwwpnsysfs, fcplunsysfs):
+        """Decide if the SCSI device with the provided SCSI attributes
+        corresponds to the zFCP device.
```

Comment:
"""Decide if the provided FCP addressing
corresponds to the path stored in the zFCP device.